### ISSUE TITLE
No warning on [[nodiscard]] instance

### ISSUE THREAD

**Issue by JVApen** [llvm/llvm-project#52686]
When a function returns a value that's no-discard, that function can be passed to construct std::function<void()> which will ignore the value and this doesn't trigger the nodiscard warning. Strangly enough MSVC does print the warning.

````cpp
// clang++ -std=c++20 -O3 -Weverything -Wno-c++98-compat -Wno-missing-prototypes t.cpp
#include <functional>

struct [[nodiscard]] A{};

auto f(std::function<void()>) -> void;

auto g() -> void
{
    f([](){ return A{}; });
}
````
Reproduction: https://compiler-explorer.com/z/5acnssWEo

**Comment by dwblaikie** [comment]
Yeah, this is an issue with clang diagnostic suppression in system headers: https://compiler-explorer.com/z/YaTfYEno7

My understanding is this is a general problem with Clang's warning/system header infrastructure - I'm not sure if there's a useful systemic fix that could use the fact that it's a template instantiation to determine whether to suppress the warning.

Generally Clang's warnings are setup well to only apply to the template pattern or the template instantiation - so if the warning only comes up when instantiating the template and the template instantiation stack includes non-system code (eg: if it's a non-template function in a system header that instantiates and causes the problem, that should be suppressed because the user has no involvement in it - but if it's a series of template instantiations that does lead from non-system code to the system code where the error is... maybe that's user-addressable? Though it's not /for sure/ - there are probably cases where still the user can't do anything about the diagnostic which is problematic)

@AaronBallman - thoughts?

**Comment by AaronBallman** [comment]
> My understanding is this is a general problem with Clang's warning/system header infrastructure - I'm not sure if there's a useful systemic fix that could use the fact that it's a template instantiation to determine whether to suppress the warning.

I agree with your analysis that this is an issue with our diagnostic suppression mechanisms for system headers.

> ...but if it's a series of template instantiations that does lead from non-system code to the system code where the error is... maybe that's user-addressable? Though it's not /for sure/ - there are probably cases where still the user can't do anything about the diagnostic which is problematic

The immediate cases that all spring to mind are situations that are under the user's control, but I share your fear that there are probably cases where the user can't do anything about the diagnostics. But there are also definitely cases where the user should be alerted to a problem but aren't because of the system header suppression mechanisms (like this bug report, or even code like https://compiler-explorer.com/z/aYYY9746E).

Due to this being related to template instantiations, there's no way currently (that I know of) to silence the *instantiation* warnings with a pragma. e.g., you can't do this to silence the warning: https://compiler-explorer.com/z/xKzjT5sG7 If we had some ability to say "silence diagnostics generated by the statements or instantiating the statements", then we could probably relax our system header diagnostic suppression around template instantiations because users would have an out for the rare (at least, I think they should be rare) situations where they can't otherwise silence the issue in the system header. However, I have no idea how easy/hard that would be to implement. I suppose another alternative would be something like `#pragma clang diagnostic ignore "-Wwhatever" in file.h:21` to silence diagnostics within other files, but I think that's super fragile for a bunch of reasons (line numbers moving, file names being wonky sometimes, etc).

**Comment by dwblaikie** [comment]
Ah, yeah, good points all - especially about diagnostic suppression. Figuring out some way to suppress these diagnostics seems key to alleviating the high cost/concern about potential false positives by taking this approach.

**Comment by JVApen** [comment]
Just wondering would `#pragma clang diagnostic ignore "-Wwhatever" in template` be an option? It might not be fine grained, though it would be rigid and could than hold for the complete template instation. So if the warning happens multiple times, it could suppress all of them.

**Comment by dwblaikie** [comment]
> Just wondering would `#pragma clang diagnostic ignore "-Wwhatever" in template` be an option? It might not be fine grained, though it would be rigid and could than hold for the complete template instation. So if the warning happens multiple times, it could suppress all of them.

The problem is the template is in a system header - but the user code that triggered the warning is in user code/non-system. So the challenge is providing some way for the user to suppress the warning. (but also probably for the template author to suppress it too - so we might have to check multiple locations and coalesce the results in some way... )

**Comment by JVApen** [comment]
I think I expressed myself badly, I was thinking of adding something like that where the template is instantiated.
So it holds for all template instantiations that follow this line (until the pop)

**Comment by AaronBallman** [comment]
That's effectively what I was suggesting in https://github.com/llvm/llvm-project/issues/52686#issuecomment-994825453. I think such a utility would be interesting to explore, but I have no idea how easy or hard it might be to implement.
